File: ebot_mani/scripts/filters.py
"""
Filters

Author: eYRC_SB_363
"""


# Import
import pcl
import numpy as np
import rospy

# Import from files
from utils import PCLPointCloudXYZ, ROSPointCloud, PCLPointCloudXYZRGB


class Filters(object):
    """
    Implementation of filters
    """
    def __init__(self):
        pass


    def voxel_filter(self, point_cloud, leaf_size, save_flag):
        """
        - It creates a 3D voxel grid over the input point cloud data.
        Then, in each voxel, all the points present will be approximated 
        with their centroid.
        - The bigger the leaf size less the information retained.
        """
        voxel_filter = point_cloud.make_voxel_grid_filter()
        voxel_filter.set_leaf_size(leaf_size, leaf_size, leaf_size)
        point_cloud = voxel_filter.filter()
        if save_flag:
            pcl.save(point_cloud, 'voxel.pcd')
            rospy.loginfo('Saved voxel')
            rospy.loginfo('size: ' + str(point_cloud.size))
            rospy.loginfo('height: ' + str(point_cloud.height))
            rospy.loginfo('width: ' + str(point_cloud.width))
            rospy.loginfo('')
 
        return point_cloud


    def passthrough_filter(self, point_cloud, axis, min_range, max_range, save_flag):
        """
        Filter out points which are outside a specific range of a specific axis
        """
        pass_filter = point_cloud.make_passthrough_filter()
        pass_filter.set_filter_field_name(axis)
        pass_filter.set_filter_limits(min_range, max_range)
        pass_filter.filter()
        if save_flag:
            pcl.save(point_cloud, 'passthrough.pcd')
            rospy.loginfo('Saved passthrough')
            rospy.loginfo('size: ' + str(point_cloud.size))
            rospy.loginfo('height: ' + str(point_cloud.height))
            rospy.loginfo('width: ' + str(point_cloud.width))
            rospy.loginfo('')
 
        return point_cloud

        
    def cropbox(self, point_cloud, translation, rotation, ranges, save_flag):
        """
        This method is similar to the pass-through filter, 
        but instead of using three pass-through filters in series, 
        you can use one CropBox filter.
        """
        clipper = point_cloud.make_cropbox()
        clipper.set_Translation(translation[0], translation[1], translation[2])
        clipper.set_Rotation(rotation[0], rotation[1], rotation[2])
        clipper.set_MinMax(
            ranges[0], ranges[1], ranges[2], ranges[3],
            ranges[4], ranges[5], ranges[6], ranges[7]
        )
        point_cloud = clipper.filter()
        if save_flag:
            pcl.save(point_cloud, 'cropbox.pcd')
            rospy.loginfo('Saved cropbox')
            rospy.loginfo('size: ' + str(point_cloud.size))
            rospy.loginfo('height: ' + str(point_cloud.height))
            rospy.loginfo('width: ' + str(point_cloud.width))
            rospy.loginfo('')

        return point_cloud


    def ransac_filter(self, point_cloud, threshold, save_flag):
        """
        RANSAC - RANdom SAmple Consensus
        Consensus - A general agreement

        - Return pointcloud without pointcloud making plane 
        """
        # Return a pcl.Segmentation object with this object set as the input-cloud
        segmenter = point_cloud.make_segmenter()
        segmenter.set_model_type(pcl.SACMODEL_PLANE)
        segmenter.set_method_type(pcl.SAC_RANSAC)
        segmenter.set_distance_threshold(threshold)

        inlier_indices, coefficients = segmenter.segment()
        # inliers = point_cloud.extract(inlier_indices, negative = False)
        point_cloud = point_cloud.extract(inlier_indices, negative = True)

        if save_flag:
            pcl.save(point_cloud, 'ransac.pcd')
            rospy.loginfo('Saved ransac')
            rospy.loginfo('size: ' + str(point_cloud.size))
            rospy.loginfo('height: ' + str(point_cloud.height))
            rospy.loginfo('width: ' + str(point_cloud.width))
            rospy.loginfo('')

        return point_cloud


    def statistical_outlier_filter(self, point_cloud, mean_k, std_dev_mul_thresh, save_flag):
        """
        This filters out the statistical noise of the scene

        - Return pointcloud without noise
        """
        noise_filter = point_cloud.make_statistical_outlier_filter()
        
        # Set the number of points (k) to use for mean distance estimation. 
        noise_filter.set_mean_k(mean_k)

        # Any point with a mean distance larger than global (mean distance+x*std_dev)
        # will be considered outlier
        noise_filter.set_std_dev_mul_thresh(std_dev_mul_thresh)

        noise_filter.set_negative(False)
        point_cloud = noise_filter.filter()
        # noise_filter.set_negative(True)
        # outliers = noise_filter.filter()

        if save_flag:
            pcl.save(point_cloud, 'statistical_outlier.pcd')
            rospy.loginfo('Saved statistical_outlier')
            rospy.loginfo('size: ' + str(point_cloud.size))
            rospy.loginfo('height: ' + str(point_cloud.height))
            rospy.loginfo('width: ' + str(point_cloud.width))
            rospy.loginfo('')

        return point_cloud


    def euclidean_cluster_extraction(self, point_cloud_xyz, cluster_tolerance, 
                                     max_cluster_size, min_cluster_size, save_flag):
        """
        Separate clusters from the pointcloud
        """
        tree = point_cloud_xyz.make_kdtree()
        ec = point_cloud_xyz.make_EuclideanClusterExtraction()
        ec.set_ClusterTolerance(0.02)
        ec.set_MinClusterSize(min_cluster_size)
        ec.set_MaxClusterSize(max_cluster_size)
        ec.set_SearchMethod(tree)
        cluster_indices = ec.Extract()
        clusters = []

        for j, indices in enumerate(cluster_indices):
            points = [0 for i in range(len(indices))]
            for i, index in enumerate(indices):
                points[i] = point_cloud_xyz[index]

            clusters.append(PCLPointCloudXYZ())
            clusters[j].from_list(points)

            if save_flag:
                pcl.save(clusters[j], 'euclidean_cluster_' + str(j) + '.pcd')
                rospy.loginfo('Saved euclidean_cluster_' + str(j))
                rospy.loginfo('size: ' + str(clusters[j].size))
                rospy.loginfo('height: ' + str(clusters[j].height))
                rospy.loginfo('width: ' + str(clusters[j].width))
                rospy.loginfo('')

        return clusters

File: object_recognition/src/filters.cpp
/*
Author: eYRC_SB_363
*/

#include <object_recognition/filters.hpp>

Filters::Filters() {};

Filters::~Filters() {};

// Voxel filter: Downsamples the pointcloud
pcl::PointCloud<PointType>::Ptr Filters::voxel_filter(
    pcl::PointCloud<PointType>::Ptr &pointcloud_ptr,
    std::vector<float> leaf_size
)
{
    std::cout << "No. of points before voxel: " << pointcloud_ptr->size() << std::endl;
    pcl::PointCloud<PointType>::Ptr filtered_cloud_ptr(new pcl::PointCloud<PointType>());
    pcl::VoxelGrid<PointType> voxel_filter;
    voxel_filter.setInputCloud(pointcloud_ptr);
    voxel_filter.setLeafSize(leaf_size[0], leaf_size[1], leaf_size[2]);
    voxel_filter.filter(*filtered_cloud_ptr);
    std::cout << "No. of points after voxel: " << filtered_cloud_ptr->size() << std::endl;

    return filtered_cloud_ptr;
}

// Cropbox filter: Remove the points outside the specified range
pcl::PointCloud<PointType>::Ptr Filters::cropbox_filter(
    pcl::PointCloud<PointType>::Ptr &pointcloud_ptr,
    Eigen::Vector4f min_range,
    Eigen::Vector4f max_range,
    Eigen::Vector3f translation,
    Eigen::Vector3f rotation
)
{
    std::cout << "No. of points before cropbox: " << pointcloud_ptr->size() << std::endl;
    pcl::PointCloud<PointType>::Ptr filtered_cloud_ptr(new pcl::PointCloud<PointType>());
    pcl::CropBox<PointType> cropbox_filter;
    cropbox_filter.setInputCloud(pointcloud_ptr);
    cropbox_filter.setMin(min_range);
    cropbox_filter.setMax(max_range);
    cropbox_filter.setTranslation(translation);
    cropbox_filter.setRotation(rotation);
    cropbox_filter.filter(*filtered_cloud_ptr);
    std::cout << "No. of points after cropbox: " << filtered_cloud_ptr->size() << std::endl;

    return filtered_cloud_ptr;
}

// RANSAC filter: Separate the pointcloud in inliers and outliers, according to the specified model
pcl::PointCloud<PointType>::Ptr Filters::ransac_filter(
    pcl::PointCloud<PointType>::Ptr &pointcloud_ptr,
    float threshold
)
{
    std::cout << "No. of points before ransac: " << pointcloud_ptr->size() << std::endl;
    pcl::PointCloud<PointType>::Ptr filtered_cloud_ptr(new pcl::PointCloud<PointType>);
    pcl::SACSegmentation<PointType> sac;
	pcl::ModelCoefficients::Ptr coefficients(new pcl::ModelCoefficients);
	pcl::PointIndices::Ptr inlier_indices_ptr(new pcl::PointIndices);
	pcl::PointCloud<PointType>::Ptr outlier_ptr(new pcl::PointCloud<PointType>);
 
	sac.setInputCloud(pointcloud_ptr);
	sac.setMethodType(pcl::SAC_RANSAC);
	sac.setModelType(pcl::SACMODEL_PLANE);
	sac.setDistanceThreshold(threshold); 
	sac.segment(*inlier_indices_ptr, *coefficients);

    pcl::ExtractIndices<PointType> extract;
    extract.setInputCloud(pointcloud_ptr);
    extract.setIndices(inlier_indices_ptr);
    extract.setNegative(true);
    extract.filter(*filtered_cloud_ptr);
    std::cout << "No. of points after ransac: " << filtered_cloud_ptr->size() << std::endl;

    return filtered_cloud_ptr;
}

// Remove the outlier points
pcl::PointCloud<PointType>::Ptr Filters::statistical_outlier_filter(
    pcl::PointCloud<PointType>::Ptr &pointcloud_ptr
)
{
    std::cout << "Statistical outlier filter isn't implemented" << "\n" << std::endl;
}

std::vector<pcl::PointCloud<PointType>::Ptr> Filters::cluster_filter(
    pcl::PointCloud<PointType>::Ptr &pointcloud_ptr,
    float tolerance,
    int min_cluster_size,
    long max_cluster_size
)
{
    std::cout << "No. of points before cluster: " << pointcloud_ptr->size() << std::endl;
    std::vector<pcl::PointCloud<PointType>::Ptr> ptrs_cluster;
    pcl::search::KdTree<PointType>::Ptr tree_ptr(new pcl::search::KdTree<PointType>);
    pcl::EuclideanClusterExtraction<PointType> ec_ext;
    std::vector<pcl::PointIndices> clusters_indices;

    tree_ptr->setInputCloud(pointcloud_ptr);
    ec_ext.setClusterTolerance(tolerance);
    ec_ext.setMinClusterSize(min_cluster_size);
    ec_ext.setMaxClusterSize(max_cluster_size);
    ec_ext.setSearchMethod(tree_ptr);
    ec_ext.setInputCloud(pointcloud_ptr);
    ec_ext.extract(clusters_indices);

    int j = 0;
    for (std::vector<pcl::PointIndices>::const_iterator it = clusters_indices.begin(); it != clusters_indices.end(); ++it)
    {   
        pcl::PointCloud<PointType>::Ptr cloud_cluster_ptr(new pcl::PointCloud<PointType>);
        for (std::vector<int>::const_iterator p_it = it->indices.begin (); p_it != it->indices.end (); ++p_it)
            cloud_cluster_ptr->push_back ((*pointcloud_ptr)[*p_it]);
        cloud_cluster_ptr->width = cloud_cluster_ptr->size();
        cloud_cluster_ptr->height = 1;
        cloud_cluster_ptr->is_dense = true;
        ptrs_cluster.push_back(cloud_cluster_ptr);
        std::cout << "No. of points in cluster " << j << ": " << ptrs_cluster[j]->size() << std::endl;
        j++;
    }

    return ptrs_cluster;

}

